### ISSUE TITLE
double-tap fires zoom events in wrong order on iOS

### ISSUE THREAD

**Issue by cambecc** [d3/d3-zoom#75]
Using d3-zoom 1.1.1, a double-tap on iOS 10.1.1 yields this zoom event sequence:
```
start, end, start, zoom, ..., zoom, end, zoom, ..., zoom
```
The expected behavior is for the last event to be `end`:
```
start, end, start, zoom, ..., zoom, end
```
To repro, double-tap on the green cavas:
```
<html lang="en">
<head><style>p {margin: 0}</style></head>
<body>
<canvas style="background: green; width: 100%; height: 200px;"></canvas>
<div id="log"></div>
<script src="https://d3js.org/d3.v4.js"></script>
<script>
    "use strict";
    var zoom = d3.zoom();
    var log = d3.select("#log");
    zoom.on("start", function() { log.append("p").text("start"); });
    zoom.on("zoom", function() { log.append("p").text("zoom"); });
    zoom.on("end", function() { log.append("p").text("end"); });
    d3.select("canvas").call(zoom);
</script>
</body>
</html>
```
~~Live repro here: https://jsfiddle.net/mtyzm1gc/~~

**EDIT**

Improved repro here: https://bl.ocks.org/cambecc/746946d403229182dda7ef0493efce99

**Comment by powersparks** [comment]
I don't think this is an issue.
Mike wants us to put tests in the bl.ocks.org or RunKit and not use jsfiddle.. since it's going to effect the outcome of the test.. so the fail may be caused by jsfiddle.
he set up this test for zoom and double tap works pretty well:
https://bl.ocks.org/mbostock/91b75091873d06add0666f35c2e6afb4

can you confirm this is still an issue?

**Comment by cambecc** [comment]
@powersparks thanks for the clarification. Here's a block demonstrating d3-zoom's event sequence: https://bl.ocks.org/cambecc/746946d403229182dda7ef0493efce99

On iOS 10.1.1, although the underlying transform has a smoothly transitioning value, the `end` event occurs in the middle of a double-tap sequence. The effects are weird if the `end` event is used to implement a closing/finishing behavior.

**Comment by mbostock** [comment]
Debugging iOS is a real pain. Are you willing to help? The first thing I would try is changing the touchstarted handler from this:

```js
if (!g.touch1) {
  g.end();
  p = select(this).on("dblclick.zoom");
  if (p) p.apply(this, arguments);
  return;
}
```

To this:

```js
if (!g.touch1) {
  if (!started) g.end();
  p = select(this).on("dblclick.zoom");
  if (p) p.apply(this, arguments);
  return;
}
```

(You could also try throwing an error if *gesture*.active is ever negative.)

Or possibly this, though I don’t think the order makes a difference:

```js
if (!g.touch1) {
  p = select(this).on("dblclick.zoom");
  if (p) p.apply(this, arguments);
  if (!started) g.end();
  return;
}
```

**Comment by cambecc** [comment]
sure! will take a look

**Comment by cambecc** [comment]
Found the cause but not sure which fix you want. The `dblclick` handler is triggered on the second `touchstart`. This creates a race condition between the zoom transition and the second `touchend` event. Leave your finger on the screen long enough for the zoom transition to finish and there's no problem. Lift it earlier, and the `end` event is emitted while the zoom transition is still running.

Perhaps `dblclick` should instead be triggered on `touchend`? This would be consistent with browsers' treatment of double-click via mouse and also the [iOS multitouch guide](https://developer.apple.com/library/content/documentation/EventHandling/Conceptual/EventHandlingiPhoneOS/multitouch_background/multitouch_background.html#//apple_ref/doc/uid/TP40009541-CH5-SW19). It would be inconsistent with D3v3, though.

**Comment by mbostock** [comment]
The zoom behavior tracts the modalities for each gesture independently. Touch, mouse, and programmatic transitions are the three supported modalities. So if you double-tap with one finger, the touch modality is present, and if that starts a zoom transition, the transition modality is also present on the same *gesture* object, making *gesture*.active === 2. Thus, even if the *touchend* event is received before the zoom transition finishes, that reduces *gesture*.active to 1, and so it should not emit an *end* event. The *end* event should not be emitted until *gesture*.active should === 0 when the transition finishes, or when the *touchend* is received if that happens *after* the transition ends.

TL;DR reference counting

**Comment by cambecc** [comment]
Ahh I see now why you suggested adding a condition around g.end() (which unfortunately did not fix the issue). Will keep investigating.

**Comment by cambecc** [comment]
Another observation: shouldn't the double-tap detector _start_ the gesture rather than end it?
```
      if (!g.touch1) {
-       g.end();
+       g.start();
        p = select(this).on("dblclick.zoom");
        if (p) p.apply(this, arguments);
        return;
      }
```
A tap gesture has just been started, after all, even if it is the second of two taps. Eventually the user will lift their finger, so a ref increment is needed somewhere for the `touchend` to consume.

With this change, the double-tap event sequence emitted by d3-zoom looks like:
```
start, end, start, zoom, ..., zoom, end
```
Unfortunately, depending on the timing of `touchend`, this event sequence also occurs:
```
start, end, start, end, start, zoom, ..., zoom, end
```

**Comment by mbostock** [comment]
My thought was that the double-tap detector needed to record both the *end* of touching (which starting at the previous *touchstart*) and the *start* of the transition (which happens inside the dblclick handler).

However, it’s possible that this is a mistake from an earlier implementation, which did not register *touchmove* and *touchend* listeners unless the *touchstart* event initiated a gesture. The current implementation always has *touchmove* and *touchend* event listeners registered, so you might be right that it’s sufficient to record the end of touch on *touchend*. In that case, the recommend change would just be:

```
      if (!g.touch1) {
-       g.end();
        p = select(this).on("dblclick.zoom");
        if (p) p.apply(this, arguments);
        return;
      }
```

But I would verify that the *touchend* listener is actually ending the gesture as expected!

**Comment by cambecc** [comment]
Initially tried doing this (removing _g.end()_), but that is not sufficient because there is still a race condition. If `touchend` fires before the end of the zoom transition, then _touchended_ closes the gesture created by the zoom transition.

More explicitly, here's the timeline showing 1) the input event, 2) active gesture (g<_instance_><_+/-ref_>), and 3) emitted event:
```
touchstart      g0+1 -> start
touchend        g0-1 -> end
touchstart      g1
transitionstart g2+1 -> start
tween                -> zoom
tween                -> zoom
touchend        g2-1 -> end
tween                -> zoom
tween                -> zoom
transitionend   g3-1
```
If the second `touchstart` instead invoked _g.start()_, then:
```
touchstart      g0+1 -> start
touchend        g0-1 -> end
touchstart      g1+1 -> start
transitionstart g1+1
tween                -> zoom
tween                -> zoom
touchend        g1-1
tween                -> zoom
tween                -> zoom
transitionend   g1-1 -> end
```
But because `touchend` can fire at any time, the output is inconsistent since this sequence sometimes results:
```
touchstart      g0+1 -> start
touchend        g0-1 -> end
touchstart      g1+1 -> start
touchend        g1-1 -> end
transitionstart g2+1 -> start
tween                -> zoom
tween                -> zoom
transitionend   g2-1 -> end
```
I propose this sequence is actually the most desirable. It's exactly the sequence emitted by mouse double-click, and mostly conforms with [apple's definition of double-tap](https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariWebContent/HandlingEvents/HandlingEvents.html#//apple_ref/doc/uid/TP40006511-SW7): two quickly successive taps, where a tap is a `[touchstart, touchend]` gesture. The zoom transition would always begin after the second touchend.

Minor benefit: a tap-drag gesture (`[touchstart, touchend, touchstart, touchmove, ...]`) would not start a zoom transition like it does today.

Will provide a merge request implementing this behavior.

**Comment by cambecc** [comment]
Looping back on this. If there's anything I can do to help verify this fix or modify it in any way, please let me know. FWIW, this fix has been running in earth.nullschool.net for the past month. So far so good.

By far the simplest fix is the one line change mentioned in my [earlier comment](https://github.com/d3/d3-zoom/issues/75#issuecomment-267133754). However, it exhibits a slight inconsistency as described.

**Comment by heallen** [comment]
I changed the tap behavior to match cambecc's #79 (Fix#75 above), and it works great. I was facing the undesired behavior of tap + drag = zoom before the fix. Thanks cambecc! Hoping this can be merged.

**Comment by mbostock** [comment]
Is this still an issue?

**Comment by cambecc** [comment]
Yes, this is still an issue. I've made a new repro for d3 v5 (which as of this comment is d3 5.9.7): https://bl.ocks.org/cambecc/9994e223ddc48242c00699c2ea5b2175. Just double tap in the green area on an iOS device and you'll see that `end` is emitted before the last `zoom`. 

(Note: the merge request associated with this issue has a conflict, and I've not kept up with the changes to this module. I suspect the same general idea still applies.)

**Comment by mbostock** [comment]
Thanks for the response. I’m (finally) working on a new release of this library, and I will be looking at this issue soon!
